### PR TITLE
[No QA] method to randomly generate 64 bit int as string

### PR DESCRIPTION
**Pull request by srikarparsi**
### Details

New Dot section of design doc: https://docs.google.com/document/d/125xZPDi7v6jmS-GKEcvcpdAwpDRPPGsq_iEOCl9RQFw/edit#bookmark=id.3ifommrba4g8

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/223137

### Tests

- [ ] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [ ] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps

- [ ] Verify that no errors appear in the JS console

### CODE REVIEW

File: src/CONST.js [hunk 1]
```diff
@@ -790,6 +790,9 @@ const CONST = {
         INVITE: 'invite',
         LEAVE_ROOM: 'leaveRoom',
     },
+
+    MAX_64BIT_TOP_HALF: 9223372036,
```
Comment (flodnv) [inline src/CONST.js:794]: Isn't 9223372036 the bottom half? 😕 Or is this name confusing (me)?

**Review by flodnv — COMMENTED** [review]
Can we add unit tests to this? 🤔

File: src/libs/NumberUtils.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import CONST from "../CONST";
+
+function rand64() {
```
Comment (arosiclair) [inline src/libs/NumberUtils.js:3]: Let's add a jsdoc here that explains that this generates a positive 64 bit number and returns it as a `string` and that this is used for generating transaction IDs. Otherwise looks good!

**Review by arosiclair — DISMISSED** [review]

File: src/libs/NumberUtils.js (hunk 2 above)
Comment (flodnv) [reply src/libs/NumberUtils.js:3]: s/transaction IDs/IDs

(we are about to reuse this for reportActionIDs)

File: src/libs/NumberUtils.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * @returns {string} randomly generated 64 bit string
+ */
+function rand64() {
+    // Max 64-bit signed:
+    // 9,223,372,036,854,775,807
+    // The top part of the max 64-bit number *+1* because we're flooring it
+    let top = Math.floor(Math.random() * CONST.MAX_64BIT_TOP_HALF + 1);
+  
+    // The bottom part of the max 64-bit number *+1* for the same reason
+    let rest = Math.floor(Math.random() * CONST.MAX_64BIT_BOTTOM_HALF + 1);
```
Comment (stitesExpensify) [inline src/libs/NumberUtils.js:14]: Let's either make this into a ternary, or an else below so that we don't have to generate a number twice for any number that doesn't start with the max first half

File: src/CONST.js (hunk 1 above)
Comment (flodnv) [reply src/CONST.js:794]: Heh, we must think of it differently, I see it as the bottom half (the bottom of the number) 😄 Left and right sounds good to me

File: src/libs/NumberUtils.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  
+    // If the top is any number but the highest one, we can actually have any value for the rest
+    if (top != CONST.​​MAX_64BIT_TOP_HALF) {
+      rest = Math.floor(Math.random() * 1000000000);
```
Comment (stitesExpensify) [inline src/libs/NumberUtils.js:18]: @flodnv @luacmartins you can double check my snake challenge knowledge, but if we want to skip the first 1000000 for reportIDs (or more for reportActionIDs) we could:

1. Change the method to rand64(minimum)
2. Update this to `Math.floor(Math.random() * (1000000000 - minimum)) + minimum`
3. Update the other bottomHalf to `Math.floor(Math.random() * (CONST.MAX_64BIT_BOTTOM_HALF - minimum) + 1) + minimum`;

Comment (stitesExpensify) [reply src/libs/NumberUtils.js:18]: And then when getting an ID for reportIDs just pass `100000000`, transactions will pass its minimum based on our current max in the DB, etc

File: src/CONST.js (hunk 1 above)
Comment (luacmartins) [reply src/CONST.js:794]: Yea, we can change it to left and right 👍

File: src/libs/NumberUtils.js (hunk 2 above)
Comment (luacmartins) [reply src/libs/NumberUtils.js:7]: Agree with JSDocs and yes, the usage will be more than just transactionIDs

File: src/libs/NumberUtils.js (hunk 4 above)
Comment (srikarparsi) [reply src/libs/NumberUtils.js:18]: Would this way work? Let's say the left half is 8223372036, if minimum is 1000000, then you wouldn't be able to reach 8223372036 099999999. 

Also, if min is 1000000000, then rest would always be 0 which should not be the case I think.

File: src/libs/NumberUtils.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import CONST from "../CONST";
+
+/**
+ * Generates a random positive 64 bit integer by randomly generating the left half and right half and concatenating them
```
Comment (luacmartins) [inline src/libs/NumberUtils.js:4]:
```suggestion
 * Generates a random positive 64 bit numeric string by randomly generating the left half and right half and concatenating them. Used to generate client-side ids.
```

File: src/libs/NumberUtils.js (hunk 4 above)
Comment (luacmartins) [reply src/libs/NumberUtils.js:18]: I actually think we don't need to worry about using a minimum.  We are not doing it for transactionIDs and we have a lot more transactions in the DB.

Since the 64bit distribution is so sparse, the 85m reports we have make almost no difference in increasing the chances of a collision.

Comment (stitesExpensify) [reply src/libs/NumberUtils.js:18]: > ? Let's say the left half is 8223372036, if minimum is 1000000, then you wouldn't be able to reach 8223372036 099999999 for example because rest is always greater than minimum which is 1000000 in this case.

Thinking through this:
1. We have left = 8223372036
2. We generate a number between 0 and 999,000,000 (so max would be 998,999,999)
3. We add 1,000,000 making the max right side 999,999,999, and min right side 1,000,000

> Also, if min is 1000000000, then rest would always be 0 which should not be the case I think.

Good point about the minimum being higher than the right side (it is for reportActionID), we will need conditional logic for that.

Comment (stitesExpensify) [reply src/libs/NumberUtils.js:18]: @luacmartins we have over a billion transactionIDs. I know that's still small compared to 9 quintillion, but it's not nothing /shrug

Comment (srikarparsi) [reply src/libs/NumberUtils.js:18]: I'm trying to come up with a good way but I can't yet, here's my thoughts. The idea is to split the string into the left half, the left of the minimum and then the rest of the right half. Then we could use Brandon's logic of subtracting the min, multiplying it by the random number and adding the min back to the right half. But the problem with this is if the minimum is so big it goes into the right half, then we have the same initial problem since we can only generate 32 bit random numbers. To fix this we could add a max minimum which is equal to CONST.MAX_64BIT_BOTTOM_HALF (right half). I thought this was hacky but I think we have to have this condition regardless even if we do it the other way. The other way I believe I read in the thread is to generate random numbers until one is greater than the min. But to do this, the min has to be below the bottom half since if it goes into the top half, then we are comparing a >32 bit number with a >32 bit number which I don't think is natively possible. I'm re-reading this and I realized it's kind of confusing so let me know if I can elaborate on any portion.

Basically, my proposal is to add a constraint that the min has to be less than the CONST.MAX_64BIT_BOTTOM_HALF and then split the right string into two parts and if the left part is all 0's then do (rand * (right half magnitude - min)) + min and concatenate all three portions.

Comment (stitesExpensify) [reply src/libs/NumberUtils.js:18]: Can't we just do:

```
if (minimum > MAX_64BIT_BOTTOM_HALF) {
   leftMinimum = minimum.toString().substring(0,minimum.toString().length - 9);
   // code I had up there but using leftMinimum instead of minimum
} else {
   // code I had up there
}
```

Comment (luacmartins) [reply src/libs/NumberUtils.js:18]: Well 1 billion is only 0.0000000108% of the whole available range. I think the chances of having a collision with or without setting a minimum would not change much 🤷 

We decided against it in this thread - https://docs.google.com/document/d/125xZPDi7v6jmS-GKEcvcpdAwpDRPPGsq_iEOCl9RQFw/edit?disco=AAAAdh4GcbM

Comment (stitesExpensify) [reply src/libs/NumberUtils.js:18]: Okay let's just not bother then. Sorry for the confusion @srikarparsi

**Review by stitesExpensify — DISMISSED** [review]

**Review by stitesExpensify — DISMISSED** [review]
LGTM! We can ignore the linter error since we know it's not getting used yet

File: src/libs/NumberUtils.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Generates a random positive 64 bit numeric string by randomly generating the left half and right half and concatenating them. Used to generate client-side ids.
+ * @returns {string} randomly generated 64 bit string
+ */
+function rand64() {
+    // Max 64-bit signed:
+    // 9,223,372,036,854,775,807
+    // The left part of the max 64-bit number *+1* because we're flooring it
+    const left = Math.floor(Math.random() * CONST.MAX_64BIT_LEFT_HALF + 1);
```
Comment (luacmartins) [inline src/libs/NumberUtils.js:11]:
```suggestion
    const left = Math.floor(Math.random() * (CONST.MAX_64BIT_LEFT_HALF + 1));
```

File: src/libs/NumberUtils.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+    // The right part of the max 64-bit number *+1* for the same reason
+    let right;
+
+    // If the top is any number but the highest one, we can actually have any value for the rest
+    if (left !== CONST.MAX_64BIT_LEFT_HALF) {
+        right = Math.floor(Math.random() * 1000000000);
+    } else {
+        right = Math.floor(Math.random() * CONST.MAX_64BIT_RIGHT_HALF + 1);
```
Comment (luacmartins) [inline src/libs/NumberUtils.js:20]:
```suggestion
        right = Math.floor(Math.random() * (CONST.MAX_64BIT_RIGHT_HALF + 1));
```

File: src/libs/NumberUtils.js (hunk 6 above)
Comment (luacmartins) [reply src/libs/NumberUtils.js:11]: We should +1 before multiplying

File: src/libs/NumberUtils.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import CONST from '../CONST';
+
+/**
+ * Generates a random positive 64 bit numeric string by randomly generating the left half and right half and concatenating them. Used to generate client-side ids.
+ * @returns {string} randomly generated 64 bit string
+ */
+function rand64() {
```
Comment (luacmartins) [inline src/libs/NumberUtils.js:7]: We should suppress the lint error for now `/* eslint-disable no-unused-vars */`

**Review by luacmartins — CHANGES_REQUESTED** [review]

File: src/libs/NumberUtils.js (hunk 6 above)
Comment (stitesExpensify) [reply src/libs/NumberUtils.js:11]: We should actually leave it how it is, or move it the other way right?  Otherwise we can get 0

`Math.floor((Math.random() * CONST.MAX_64BIT_LEFT_HALF) + 1);`

Comment (luacmartins) [reply src/libs/NumberUtils.js:11]: Ah that's right! I guess I failed the snake challenge 😅 FWIW Tyler's [original solution](https://expensify.slack.com/archives/C02HWMSMZEC/p1659726344024529?thread_ts=1659032751.555699&cid=C02HWMSMZEC) had the same mistake

Comment (luacmartins) [reply src/libs/NumberUtils.js:11]: I edited the suggestion above to `Math.floor((Math.random() * CONST.MAX_64BIT_LEFT_HALF) + 1);`

**Comment by luacmartins** [comment]
@flodnv I tried to think of some unit tests for this, but couldn't come up with anything useful. Did you have anything in mind?

**Comment by flodnv** [comment]
We could at least verify that the generated number is in the range we expect it to be (and is a string)

**Comment by luacmartins** [comment]
Ooh I thought it would be tricky to test the range, but string comparison seems to work!

Screen Shot 2022-08-19 at 9 44 18 AM

File: src/libs/NumberUtils.js (hunk 6 above)
Comment (srikarparsi) [reply src/libs/NumberUtils.js:11]: do we need the parentheses? I think it does the same operation regardless but if it helps for clarity I can commit the suggestion.

Comment (luacmartins) [reply src/libs/NumberUtils.js:11]: I see this warning:
Screen Shot 2022-08-19 at 9 48 45 AM

**Comment by srikarparsi** [comment]
I think the lowest number we can get with this function is 1,000,000,000. I don't think it matters that much since 64 bits is very very big but I don't know if this was discussed so I wanted to bring it up.

File: src/libs/NumberUtils.js (hunk 8 above)
Comment (srikarparsi) [reply src/libs/NumberUtils.js:7]: I was thinking about this but whoever uses the function later would have to remember to take out the eslint that we would add. I can add it though if you think it's better to do it this way.

Comment (luacmartins) [reply src/libs/NumberUtils.js:7]: Yea, I think we should add it, otherwise tests would break for others.

Comment (luacmartins) [reply src/libs/NumberUtils.js:7]: Created the issue here. I'll update the link once we merge this PR.

**Comment by luacmartins** [comment]
Good point! I think we should add first and then multiply on the left part, so the range is 0-MAX_64BIT_LEFT_HALF. For the right part we could multiply and then add so the range is 1-MAX_64BIT_RIGHT_HALF

**Comment by luacmartins** [comment]
@srikarparsi can we add unit tests as Flo suggested?

**Comment by JmillsExpensify** [comment]
@srikarparsi I wanted to check in on this PR? We've had some great process thus far, so let's do everything we can to keep that same momentum this week. When do you think it'll be ready for re-review?

**Comment by srikarparsi** [comment]
I'll add unit tests, clean up the warnings and change the bounds by the end of today :)

File: src/libs/NumberUtils.js (hunk 6 above)
Comment (tylerkaraszewski) [reply src/libs/NumberUtils.js:11]: A result of 0 should be valid here. Particularly for either `left` or `right`, so long as both aren't 0.

```
const left = Math.floor(Math.random() * (CONST.MAX_64BIT_LEFT_HALF + 1));
```

Is correct, it will give a result between 0 and `MAX_64BIT_LEFT_HALF` (9223372036) inclusive of both those numbers. The same pattern is true below for `right`.

If we're concerned about getting `0`, we can check `if (left === 0 and right === 0)` and if so, change `right` to `1`. However, the chance of this happening is orders of magnitude less than this number colliding with another ID in the DB, so we should be able to ignore it, really.

**Review by tylerkaraszewski — CHANGES_REQUESTED** [review]

**Comment by luacmartins** [comment]
Thank you @tylerkaraszewski! I agree, let's make those changes @srikarparsi!

**Comment by srikarparsi** [comment]
yup @luacmartins! I fixed the lint errors and the bounds so that it allows for 0 in the right or left half but had a question about the test. Is this the right location to add a new file that tests NumberUtils.js https://github.com/Expensify/App/tree/srikar-64bitRandom/tests/unit Also, I believe this is the logic we came up with but I couldn't figure out how to pass or fail the test based on this as none of the unit tests I looked through in the tests/unit had this structure. Most of the other test code I looked at as examples used the [describe, test and expect methods](https://github.com/Expensify/App/blob/srikar-64bitRandom/tests/unit/GitUtilsTest.js#L43.). so I wanted to ask if there was somewhere I can see how to use those methods and if that is the right way to go about writing this unit test.

Screen Shot 2022-08-23 at 3 45 42 AM

**Review by tylerkaraszewski — DISMISSED** [review]

**Comment by luacmartins** [comment]
@srikarparsi yes, that's the correct path for the tests. As for the test themselves, I'd do something like the code below:

```
describe('libs/NumberUtils', () => {
  it('should generate a random 64-bit numeric string', () => {
    const id = NumberUtils.rand64();
    expect(id).toBe(...);
    ...
  });
})

File: tests/unit/NumberUtilsTest.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+const NumberUtils = require('../../src/libs/NumberUtils');
+
+const result = NumberUtils.rand64();
+
+const max64 = 9223372036854775807;
+
+describe('libs/NumberUtils', () => {
+    it('should generate a random 64-bit numeric string', () => {
+        const id = NumberUtils.rand64();
+        expect(isNaN(result)).toBe(false);
```
Comment (luacmartins) [inline tests/unit/NumberUtilsTest.js:10]: I think this is what you are looking for
```suggestion
        expect(isNaN(id));
```

File: tests/unit/NumberUtilsTest.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+        expect(result > max64).toBe(false);
```
Comment (luacmartins) [inline tests/unit/NumberUtilsTest.js:11]: We should check if `.toBeGreaterOrEqual` works with strings.

```suggestion
        expect(id).toBeGreaterOrEqual(max64);
```

**Review by luacmartins — CHANGES_REQUESTED** [review]

File: tests/unit/NumberUtilsTest.js (hunk 10 above)
Comment (luacmartins) [reply tests/unit/NumberUtilsTest.js:11]: It seems like we can only pass number or bigint - https://jestjs.io/docs/expect#tobegreaterthanorequalnumber--bigint

```suggestion
        expect(result >= max64).toBe(true);
```

File: src/libs/NumberUtils.js (hunk 8 above)
Comment (stitesExpensify) [reply src/libs/NumberUtils.js:8]: @luacmartins there is no link on your "here" 😄

Comment (luacmartins) [reply src/libs/NumberUtils.js:8]: Classic. Edited and added the link. Thanks!

File: tests/unit/NumberUtilsTest.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+const NumberUtils = require('../../src/libs/NumberUtils');
+
+const result = NumberUtils.rand64();
+
+const max64 = 9223372036854775807;
```
Comment (tylerkaraszewski) [inline tests/unit/NumberUtilsTest.js:5]: How does this work? As far as I know, this will result in a floating point number with a loss of precision.

File: tests/unit/NumberUtilsTest.js (hunk 10 above)
Comment (tylerkaraszewski) [reply tests/unit/NumberUtilsTest.js:11]: `result` (or `id`, which one are we using?) should be *less* than the max 64-bit number. But this is an inadequate test case (especially if `max64` is wrong). We should probably check it's >0 as well.

**Review by tylerkaraszewski — CHANGES_REQUESTED** [review]

File: tests/unit/NumberUtilsTest.js (hunk 11 above)
Comment (stitesExpensify) [reply tests/unit/NumberUtilsTest.js:5]: Yeah, I think this needs to be a BigInt

File: tests/unit/NumberUtilsTest.js (hunk 9 above)
Comment (roryabraham) [inline tests/unit/NumberUtilsTest.js:10]:
```suggestion
        expect(Number.isNaN(result)).toBe(false);
```

File: tests/unit/NumberUtilsTest.js (hunk 11 above)
Comment (luacmartins) [reply tests/unit/NumberUtilsTest.js:5]: it should be a string, no? since we are generating a numeric string id

File: tests/unit/NumberUtilsTest.js (hunk 10 above)
Comment (luacmartins) [reply tests/unit/NumberUtilsTest.js:11]: we declared `const id = NumberUtils.rand64();` above, so it should be id and yes the comparison should be `id >= 0 && id <= max64)`

**Comment by stitesExpensify** [comment]
cc @tylerkaraszewski @roryabraham

File: tests/unit/NumberUtilsTest.js [hunk 12]
```diff
@@ -1,14 +1,13 @@
...
-const max64 = 9223372036854775807;
-
 describe('libs/NumberUtils', () => {
     it('should generate a random 64-bit numeric string', () => {
         const id = NumberUtils.rand64();
-        expect(isNaN(result)).toBe(false);
-        expect(result > max64).toBe(false);
+        expect(typeof id).toBe('string');
+        // eslint-disable-next-line no-undef
+        expect(BigInt(id)).toBeLessThanOrEqual(BigInt(9223372036854775807));
```
Comment (roryabraham) [inline tests/unit/NumberUtilsTest.js:8]: Using BigInt here is safe because BigInt is defined in Node and we are using the node environment for jest?

**Review by tylerkaraszewski — APPROVED** [review]

File: tests/unit/NumberUtilsTest.js (hunk 12 above)
Comment (flodnv) [reply tests/unit/NumberUtilsTest.js:8]: Can we not use string comparison?

**Review by roryabraham — APPROVED** [review]
LGTM, I'm not worried about using BigInt in tests. If it passes locally and passes in GitHub Actions, then we're good.

File: tests/unit/NumberUtilsTest.js (hunk 12 above)
Comment (stitesExpensify) [reply tests/unit/NumberUtilsTest.js:8]: We could if we want to, this just _feels_ better IMO because it makes them numbers instead of JS black magic with strings

Comment (flodnv) [reply tests/unit/NumberUtilsTest.js:8]: Black magic? All programming languages (that we use at Expensify) do string comparisons 😕 🤷 NAB

**Review by luacmartins — APPROVED** [review]

**Comment by ctkochan22** [comment]
hmm this was breaking when I was running the desktop locally. Could be me though?
```
App [main] :> npm run desktop

> new.expensify@1.1.89-2 desktop
> scripts/set-pusher-suffix.sh && node desktop/start.js

Using PUSHER_DEV_SUFFIX from ../Web-Expensify/_config.local.php
Found suffix: 69db2c8c04e54064a2d1ca48a515d180
Updating .env
[Renderer] ℹ ｢wds｣: Project is running at http://localhost:8080/
[Renderer] ℹ ｢wds｣: webpack output is served from /
[Renderer] ℹ ｢wds｣: Content not from webpack is served from /Users/kosuke/Expensidev/App/dist
[Renderer] ℹ ｢wds｣: 404s will fallback to /index.html
[Main] Hash: 93c0b04638346214ccf0
[Main] Version: webpack 4.46.0
[Main] Time: 530ms
[Main] Built at: 08/24/2022 4:04:49 PM
[Main]            Asset      Size         Chunks             Chunk Names
[Main] contextBridge.js   306 KiB  contextBridge  [emitted]  contextBridge
[Main]          main.js  2.61 MiB           main  [emitted]  main
[Main] Entrypoint main = main.js
[Main] Entrypoint contextBridge = contextBridge.js
[Main] [./desktop/ELECTRON_EVENTS.js] 356 bytes {main} {contextBridge} [built]
[Main] [./desktop/contextBridge.js] 2.39 KiB {contextBridge} [built]
[Main] [./desktop/main.js] 13.7 KiB {main} [built]
[Main] [./node_modules/webpack/buildin/module.js] (webpack)/buildin/module.js 497 bytes {main} [built]
[Main] [./src/CONFIG.js] 3.37 KiB {main} [built]
[Main] [./src/CONST.js] 416 bytes {main} [built] [failed] [1 error]
[Main] [./src/libs/Url.js] 321 bytes {main} [built]
[Main] [./src/libs/checkForUpdates.js] 630 bytes {main} [built]
[Main] [./src/libs/getPlatform/index.desktop.js] 111 bytes {main} [built]
[Main] [electron] external "electron" 42 bytes {main} {contextBridge} [built]
[Main] [electron-context-menu] external "electron-context-menu" 42 bytes {main} [built]
[Main] [electron-log] external "electron-log" 42 bytes {main} [built]
[Main] [electron-serve] external "electron-serve" 42 bytes {main} [built]
[Main] [electron-updater] external "electron-updater" 42 bytes {main} [built]
[Main]     + 457 hidden modules
[Main]
[Main] ERROR in ./src/CONST.js 793:26
[Main] Module parse failed: Identifier directly after number (793:26)
[Main] You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
[Main] |         LEAVE_ROOM: 'leaveRoom',
[Main] |     },
[Main] >     MAX_64BIT_LEFT_HALF: 9_223_372_036,
[Main] |     MAX_64BIT_RIGHT_HALF: 854_775_807,
[Main] |     IOS_KEYBOARD_SPACE_OFFSET: -30,
[Main]  @ ./src/CONFIG.js 6:0-28 9:53-58 23:20-25
[Main]  @ ./desktop/main.js
[Main] webpack --config config/webpack/webpack.desktop.js --config-name desktop-main --mode=development && wait-port localhost:8080 && electron desktop/dist/main.js exited with code 2
--> Sending SIGTERM to other processes..
^C^[[B^[[A[Renderer] webpack-dev-server --config config/webpack/webpack.dev.js --port 8080 --env.platform desktop exited with code 0
```

**Comment by luacmartins** [comment]
@ctkochan22 there's a thread [here](https://expensify.slack.com/archives/C03TQ48KC/p1661393972628819) if you want to follow
